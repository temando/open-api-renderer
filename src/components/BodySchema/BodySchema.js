import React, { Component } from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Property from '../Property/Property'
import { styles } from './BodySchema.styles'

@styles
export default class BodySchema extends Component {
  constructor (props) {
    super(props)

    this.onClick = this.onClick.bind(this)

    let expandedProps = []
    const {properties, depthToExpand} = this.props

    if (depthToExpand > 0) {
      expandedProps = properties.map(property => property.name)
    }

    this.state = {
      expandedProps
    }
  }

  render () {
    const { properties, additionalProperties, styleVariation, classes, hasTabs, depthToExpand } = this.props

    if (!properties) {
      return null
    }

    const { expandedProps } = this.state
    const hasAdditionalProperties = additionalProperties !== undefined

    return (
      <table className={classNames(classes.bodySchema, classes[styleVariation], {
        [classes.hasTabs]: hasTabs
      })}>
        <tbody>
          {properties.map((property, i) => {
            // We will never be last if there is additionalProperties to follow
            const isLast = hasAdditionalProperties ? false : (properties.length === i + 1)
            const hasProps = property.properties !== undefined
            const hasAdditional = property.additionalProperties !== undefined

            // Handle no properties and no additional properties
            if (!hasProps && !hasAdditional) {
              return this.renderPropertyRow(property, isLast)
            }

            // Handle additionalProperties only
            if (!hasProps && hasAdditional) {
              return this.renderProperties(property.additionalProperties, expandedProps, isLast, hasAdditional, depthToExpand)
            }

            // Handle properties and additional properties
            return this.renderProperties(property, expandedProps, isLast, hasAdditional, depthToExpand)
          })}
          {additionalProperties && this.renderProperties(additionalProperties, expandedProps, true, true)}
        </tbody>
      </table>
    )
  }

  /**
   * Renders a property, with additional smarts to know if the property itself
   * has a subset (because it's of type array or object). If the property is
   * expanded, then it's subset will be rendered as well.
   *
   * @see renderSubsetProperties
   * @param {Object} property
   * @param {Array} expandedProps
   * @param {Boolean} isLast
   * @param {Boolean} hasAdditional
   * @param {Number} depthToExpand
   */
  renderProperties (property, expandedProps = [], isLast = false, hasAdditional = false, depthToExpand) {
    const isPropertyArray = property.type && property.type.includes('array')
    const isPropertyObject = property.type && property.type.includes('object')

    if (isPropertyArray || isPropertyObject) {
      if (expandedProps.includes(property.name)) {
        return [
          this.renderPropertyRow(property, isLast, true, true),
          this.renderSubsetProperties(property, isPropertyArray, depthToExpand - 1)
        ]
      }

      return this.renderPropertyRow(property, isLast, false, true)
    }

    return this.renderPropertyRow(property, isLast, false, false, hasAdditional)
  }

  /**
   * Invokes the Property component, asking it to render the table row. Has a number
   * of parameters that merely help in turn styling aids on/off
   *
   * @param {Object} property
   * @param {Boolean} isLast
   * @param {Boolean} isOpen
   * @param {Boolean} hasSubset
   * @param {Boolean} isAdditionalProperties
   */
  renderPropertyRow (property, isLast, isOpen = false, hasSubset = false, isAdditionalProperties = false) {
    return (
      <Property
        key={property.name}
        name={property.name}
        type={property.type}
        subtype={property.subtype}
        description={property.description}
        enumValues={property.enum}
        defaultValue={property.defaultValue}
        constraints={property.constraints}
        onClick={hasSubset ? this.onClick : undefined}
        isRequired={property.required}
        isOpen={isOpen}
        isLast={isLast}
        isAdditional={isAdditionalProperties}
      />
    )
  }

  /**
   * Responsible for rendering array properties. This is usually only invoked
   * if the parent property has been clicked on and the user wants to see the
   * definition.
   *
   * @param {Object} property
   * @param {Boolean} isArray
   * @param {Number} depthToExpand
   */
  renderSubsetProperties (property, isArray = false, depthToExpand) {
    const { styleVariation, classes } = this.props
    const nextStyleVariation = (styleVariation === 'even') ? 'odd' : 'even'

    return (
      <tr className={classes.subset}>
        <td colSpan='100'>
          {isArray && <div className={classes.array}>Array [</div>}
          <BodySchema
            {...this.props}
            key={`${property.name}-properties`}
            properties={property.properties}
            additionalProperties={property.additionalProperties}
            styleVariation={nextStyleVariation}
            depthToExpand={depthToExpand}
          />
          {isArray && <div className={classes.array}>]</div>}
        </td>
      </tr>
    )
  }

  /**
   * Responsible for updating the state of all properties that
   * have been expanded by the user.
   *
   * @param {string} propertyName
   */
  onClick (propertyName) {
    const { expandedProps } = this.state

    if (expandedProps.includes(propertyName)) {
      const newExpanded = expandedProps.filter((prop) => prop !== propertyName)
      this.setState({ expandedProps: newExpanded })
    } else {
      this.setState({ expandedProps: [...expandedProps, propertyName] })
    }
  }
}

BodySchema.propTypes = {
  additionalProperties: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.object
  ]),
  properties: PropTypes.array,
  styleVariation: PropTypes.oneOf([
    'odd',
    'even'
  ]),
  classes: PropTypes.object,
  hasTabs: PropTypes.bool,
  depthToExpand: PropTypes.number
}
