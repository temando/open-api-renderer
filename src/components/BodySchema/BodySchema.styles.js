import { styles as property } from '../Property/Property.styles'

export const styles = theme => ({
  'bodySchema': {
    width: '100%',
    borderSpacing: '0',
    padding: '0 4rem 0 1rem',
    boxSizing: 'border-box',
    borderRadius: '1rem',
    fontSize: `${theme.sizes.text}`,

    '& td': {
      padding: '.5rem 1rem'
    }
  },

  'even': {
    backgroundColor: `${theme.backgrounds.default}`
  },
  'odd': {
    backgroundColor: `${theme.backgrounds.schema}`
  },

  'subset': {
    [`& + .${property(theme).property}:not(.${property(theme).last})`]: {
      [`& .${property(theme).name}::after`]: {
        content: '""',
        borderLeft: `1px solid ${theme.borders.default}`,
        position: 'absolute',
        bottom: '0',
        left: '1rem',
        height: '100%'
      }
    },

    [`& > td`]: {
      position: 'relative'
    },

    [`& > td::before`]: {
      content: '""',
      borderLeft: `1px solid ${theme.borders.default}`,
      padding: '0',
      position: 'absolute',
      top: 0,
      left: '1rem',
      bottom: 0
    },

    [`& > td > table`]: {
      marginLeft: '1rem'
    },

    [`.${property(theme).last} + & > td`]: {
      paddingTop: 0
    },

    [`.${property(theme).last} + & > td::before`]: {
      borderLeft: 'none'
    }
  },

  'array': {
    fontFamily: 'monospace',
    fontSize: '1.6rem',
    padding: '.5rem 1rem'
  },

  'hasTabs': {
    borderTopLeftRadius: 0
  },

  'isAdditional': {
    textAlign: 'left',
    fontStyle: 'italic',
    fontSize: `${theme.sizes.smaller}`,
    paddingBottom: '0.5rem'
  },

  [`@media (max-width: ${theme.sizes.breakpoint})`]: {
    'bodySchema': {
      paddingRight: '2rem'
    }
  }
})
