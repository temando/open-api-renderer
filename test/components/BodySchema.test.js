import React from 'react'
import BodySchema from './../../src/components/BodySchema/BodySchema'
import renderer from 'react-test-renderer'

describe('<BodySchema />', () => {
  const keyProperty = {
    name: 'key',
    required: false,
    type: [ 'string' ]
  }
  const additionalProperty = {
    name: 'additionalProperties',
    required: false,
    type: [ 'integer' ],
    constraints: {
      format: 'int32'
    }
  }

  it('can render a basic BodySchema', () => {
    const tree = renderer.create(
      <BodySchema styleVariation='even' properties={[keyProperty]} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a basic BodySchema with additional properties', () => {
    const properties = [
      {
        ...keyProperty,
        additionalProperties: additionalProperty
      }
    ]
    const tree = renderer.create(
      <BodySchema styleVariation='even' properties={properties} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a BodySchema with a subset', () => {
    const properties = [
      {
        name: 'customAttributes',
        required: true,
        type: [ 'array' ],
        properties: [
          keyProperty
        ]
      }
    ]
    const tree = renderer.create(
      <BodySchema styleVariation='even' properties={properties} />
    )

    expect(tree).toMatchSnapshot()

    // Simulate expanding customAttributes
    tree.toJSON().children[0].children[0].props.onClick('customAttributes')

    expect(tree).toMatchSnapshot()

    // Simulate collapsing customAttributes
    tree.toJSON().children[0].children[0].props.onClick('customAttributes')

    expect(tree).toMatchSnapshot()
  })

  it('can render a BodySchema with a subset and no additional properties', () => {
    const properties = [
      {
        additionalProperties: false,
        name: 'customAttributes',
        required: true,
        type: [ 'array' ],
        properties: [
          keyProperty
        ]
      }
    ]
    const tree = renderer.create(
      <BodySchema styleVariation='even' properties={properties} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a BodySchema with a subset and additional properties', () => {
    const properties = [
      {
        additionalProperties: additionalProperty,
        name: 'customAttributes',
        required: true,
        type: [ 'array' ],
        properties: [
          keyProperty
        ]
      }
    ]
    const tree = renderer.create(
      <BodySchema styleVariation='even' properties={properties} />
    )

    expect(tree).toMatchSnapshot()
  })
})
