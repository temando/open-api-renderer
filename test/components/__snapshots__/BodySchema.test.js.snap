// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BodySchema /> can render a BodySchema with a subset 1`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span>
          customAttributes
        </span>
        <div
          className=""
        >
          <img
            alt=""
            src="test-file-stub"
            title="arrow"
          />
        </div>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          array
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
          >
            required
          </span>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<BodySchema /> can render a BodySchema with a subset 2`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span>
          customAttributes
        </span>
        <div
          className=""
        >
          <img
            alt=""
            src="test-file-stub"
            title="arrow"
          />
        </div>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          array
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
          >
            required
          </span>
        </span>
      </td>
    </tr>
    <tr
      className={undefined}
    >
      <td
        colSpan="100"
      >
        <div
          className={undefined}
        >
          Array [
        </div>
        <table
          className=""
        >
          <tbody>
            <tr
              className="undefined"
              onClick={[Function]}
            >
              <td
                className=""
              >
                <span>
                  key
                </span>
              </td>
              <td
                className={undefined}
              >
                <span
                  className={undefined}
                >
                  string
                </span>
                <span
                  className={undefined}
                />
              </td>
            </tr>
          </tbody>
        </table>
        <div
          className={undefined}
        >
          ]
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<BodySchema /> can render a BodySchema with a subset 3`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span>
          customAttributes
        </span>
        <div
          className=""
        >
          <img
            alt=""
            src="test-file-stub"
            title="arrow"
          />
        </div>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          array
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
          >
            required
          </span>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<BodySchema /> can render a BodySchema with a subset and additional properties 1`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span>
          customAttributes
        </span>
        <div
          className=""
        >
          <img
            alt=""
            src="test-file-stub"
            title="arrow"
          />
        </div>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          array
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
          >
            required
          </span>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<BodySchema /> can render a BodySchema with a subset and no additional properties 1`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span>
          customAttributes
        </span>
        <div
          className=""
        >
          <img
            alt=""
            src="test-file-stub"
            title="arrow"
          />
        </div>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          array
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
          >
            required
          </span>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<BodySchema /> can render a basic BodySchema 1`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span>
          key
        </span>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          string
        </span>
        <span
          className={undefined}
        />
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<BodySchema /> can render a basic BodySchema with additional properties 1`] = `
<table
  className=""
>
  <tbody>
    <tr
      className="undefined"
      onClick={[Function]}
    >
      <td
        className="undefined"
      >
        <span>
          additionalProperties
        </span>
      </td>
      <td
        className={undefined}
      >
        <span
          className={undefined}
        >
          integer
          <span
            className={undefined}
          >
            &lt;
            int32
            &gt;
          </span>
        </span>
        <span
          className={undefined}
        />
      </td>
    </tr>
  </tbody>
</table>
`;
